crates/status/src/impl/bitstring_status_list/mod.rs

```diff
@@ -334,23 +349,155 @@ impl BitString {
             return None;
         }
 
+        self.inner.get(self.status_size, index)
+    }
+
+    /// Push a new value into the bit-string.
+    ///
+    /// Returns the index of the newly inserted value in the list,
+    /// or an error if the value is too large w.r.t. `status_size`.
+    pub fn push(&mut self, value: u8) -> Result<usize, Overflow> {
+        let masked_value = value & self.status_size.mask();
+        if masked_value != value {
+            return Err(Overflow::Value(value));
+        }
+
+        let index = self.len;
         let offset = self.status_size.offset_of(index);
+
         let (high_shift, low_shift) = offset.left_shift(self.status_size);
 
-        Some(self.get_at(offset.byte, high_shift, low_shift))
+        if offset.byte == self.inner.0.len() {
+            self.inner
+                .0
+                .push(masked_value.overflowing_signed_shl(high_shift).0);
+        } else {
+            self.inner.0[offset.byte] |= masked_value.overflowing_signed_shl(high_shift).0
+        }
+
+        if let Some(low_shift) = low_shift {
+            self.inner.0.push(masked_value.overflowing_shl(low_shift).0);
+        }
+
+        self.len += 1;
+        Ok(index)
+    }
+
+    /// Sets the value at the given index.
+    ///
+    /// Returns the previous value, or an `Overflow` error if either the index
+    /// is out of bounds or the value is too large.
+    pub fn set(&mut self, index: usize, value: u8) -> Result<u8, Overflow> {
+        if index >= self.len {
+            return Err(Overflow::Index(index));
+        }
+
+        self.inner.set(self.status_size, index, value)
+    }
+
+    /// Returns an iterator over all the statuses stored in this bit-string.
+    pub fn iter(&self) -> BitStringIter {
+        self.inner.iter(self.status_size)
+    }
+
+    /// Encodes the bit-string.
+    pub fn encode(&self) -> EncodedList {
+        self.inner.encode()
+    }
+
+    pub fn into_unsized(self) -> BitString {
+        self.inner
+    }
+}
+
+/// Bit-string as defined by the W3C Bitstring Status List specification.
+///
+/// Bits are indexed from most significant to least significant.
+/// ```text
+/// | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | ... | n-8 | n-7 | n-6 | n-5 | n-4 | n-3 | n-2 | n-1 |
+/// | byte 0                        | ... | byte k-1                                      |
+/// ```
+///
+/// See: <https://www.w3.org/TR/vc-bitstring-status-list/#bitstring-encoding>
+#[derive(Debug, Default, Clone)]
+pub struct BitString(Vec<u8>);
+
+impl BitString {
+    /// Creates a new empty bit-string.
+    pub fn new() -> Self {
+        Self::default()
     }
 
-    fn get_at(&self, byte_offset: usize, high_shift: i32, low_shift: Option<u32>) -> u8 {
+    /// Creates a new bit-string of the given length, using `f` to initialize
+    /// every status.
+    ///
+    /// The `f` function is called with the index of the initialized status.
+    pub fn new_with(
```

Comment:
wdyt of restricting the public API of `BitString` as much as possible & requiring interfacing with it through `SizedBitString` to make usage safer and avoid needing to pass in the StatusSize each time? And then perhaps adding a `TryFrom` impl from `BitString` -> `SizedBitString` so that it's possible to go from one to the other, and at least avoids needing to specify size in a bunch of places 🤔 And maybe something similar is possible for `SizedStatusList` & `BitstringStatusList`. I'm wondering if that'll make it possible to both 1) have the claim implementations conform to the spec while still 2) providing some safer, common-sense ways of interacting with the inner data structures. What are your thoughts?

Reply:
`SizedBitString` owns the string, which can be annoying if one wants to access a bit-string reference. I don't mind providing a lower-level API, even if it allows logical errors. I've added a comment for `BitString` pointing to `SizedBitString`.